JBPM-6488 Stunner copy cut paste selection

tiagodolphine opened the pull request:
Now the copy, cut paste features are available for single node selection including containers with children.
More details on https://issues.jboss.org/browse/JBPM-6488

Summary:
- Toolbar:  mew items on workbench and standalone 
- Keyboard shortcuts
Ctrl + C : copy 
Ctrl + X : cut
Ctrl + V : paste
- Properties cloning
- Node and children cloning
- Parent selection when pasting
- Code refactor on the Morph to adopt clone logic

Can you please help to review and give your feedback and suggestions.
@romartin 
@hasys 
@manstis 

Thank you!

jomarko commented:
jenkins execute full downstream build

jomarko commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/definition/clone/DeepCloneProcess.java:
```diff
@@ -0,0 +1,78 @@
+        }
+        return Objects.nonNull(value) &&
+                ((value instanceof String) || (classUtils.isPrimitiveClass(value.getClass())));
+    }
```
This method needs to enhance. I mean If the first `if` is true, then the code bellow is always false. Also I think `System.out.print("null");` should be replaced with some log.

tiagodolphine replied:
My mistake, this was just a test, I removed it. Thanks !

jomarko commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/util/ClassUtils.java:
```diff
@@ -0,0 +1,56 @@
+        initialized = true;
+    }
+
+    public boolean isPrimitiveClass(Class<?> type) {
```
Can't be this implemented using [commons-lang](http://commons.apache.org/proper/commons-lang/javadocs/api-3.6/org/apache/commons/lang3/ClassUtils.html#isPrimitiveWrapper-java.lang.Class-)?

tiagodolphine replied:
Well, using the commons-lang was my first try but since this is a client code with GWT, the java reflection doesn't work, probably this cause the error when running so I decided to implement it myself keeping it as simple as possible.

tiagodolphine commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-backend-common/src/main/java/org/kie/workbench/common/stunner/core/backend/definition/adapter/annotation/RuntimePropertyAdapter.java:
```diff
@@ -211,7 +211,7 @@ public void setValue(final T property,
         if (null != property) {
             if (isReadOnly(property)) {
-                throw new RuntimeException( "Cannot set new value for property [" + getId( property ) + "] as it is read only! " );
+                //throw new RuntimeException( "Cannot set new value for property [" + getId( property ) + "] as it is read only! " );
```
Done, I commented and forgot this point, thanks.

tiagodolphine commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/session/command/impl/CopySelectionSessionCommand.java:
```diff
@@ -0,0 +1,99 @@
+
+    private void handleCtrlC(Key[] keys) {
+        if (doKeysMatch(keys, Key.CONTROL, Key.C)) {
+            this.execute(newDefaultCallback("Error while trying to copy selected items. Message="));
```
Done.

tiagodolphine commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/session/command/impl/CutSelectionSessionCommand.java:
```diff
@@ -0,0 +1,81 @@
+
+            @Override
+            public void onError(V error) {
+                callback.onError(error);
```
I will talk to @romartin but the `callback.onError(V)` is strange the way it was implemented since the method that is receiving the callback doesn't know the V type. For now I will skip this callback.onError and keep the log on a consistent way.

tiagodolphine commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/session/command/impl/PasteSelectionSessionCommand.java:
```diff
@@ -0,0 +1,215 @@
+
+            if (CommandUtils.isError(result)) {
+                LOGGER.severe("Error pasting selection " + CommandUtils.toList(result.getViolations()));
+                callback.onError(null);
```
The way each command is executed is different from each other, the copy and cut are only on client  side, this mean none modification on graph is performed, but the paste in fact execute the clone operation, a command on the graph. So the way the error are handled my be quite different, but I tried to keep is more consistent as possible on this PR update.

jomarko commented:
@tiagodolphine thanks for your answers and updates, they seems good to me, however, I would like to ask you for double check if error callbacks are consistent now in `Copy/Cut/PasteSelectionSessionCommand` classes. It seems to me that two of them just logs and one of them also calls `onError`.

tiagodolphine commented:
Jenkins please retest this

tiagodolphine commented:
Hi @jomarko, I talked to @romartin about the callback.onError , and this part should be refactored in the future, because working with generics in that point is complicated, the caller expects one type of object that could be anything and the method cannot know how to handle, for now we will just log errors and refactor this part in other Jira. @romartin do you have any comment on this?
Thanks.

tiagodolphine commented:
@jomarko @hasys  I'm working on the issues pointed by @hasys ASAP I will update the PR answering the comments.
Thanks

tiagodolphine commented:
Jenkins please retest this

tiagodolphine commented:
Hi @hasys , thank you for your comments, they were really important.
The PR was updated with fixes and improvements.

Follow are the answer to your comments https://github.com/kiegroup/kie-wb-common/pull/1286#pullrequestreview-75765046:

- There are some failing unit tests in Stunner and DMN editor. - **OK**

- When I cut and paste subprocess with children - only subprocess appear. Copy and paste - works good, both subprocess and children appeared. - **OK**

- Attached nodes doesn't copied (not after copy, nor after cut). - **OK**

- Properties panel didn't clear after node cut. Expected - Process properties will be displayed. - **OK**

- Sometimes pasted subprocess’s sequence flows displayed under subprocess. See pic 1. - **Cannot reproduce**

- When I click fast Ctrl+C, Ctrl+V I see this error pic 2. I need to wait 1-2 seconds between Ctrl+C, Ctrl+V for successful copy paste. Sometimes I see it even after several seconds. If I click on the empty space on canvas, Ctrl+V it work, but node is still added above copied see pic 3. - **OK**

- It looks like copy and paste feature has memory leaks. When I copy and paste, move and delete single nodes / subprocesses memory usage is growing fast. I achieved 3+ GB of memory usage by browser tab and my laptop starts to freeze. I saved Business Process and reopened it. But it didn’t help, Firefox starts to consume even more memory.
I closed browser tab and reopen entire business central. After this action memory consuming reduced to 420 MB.
When I model Business Process without Copy and Paste feature, memory usage increased slowly and decreased with time automatically (maybe keep some addition memory against freshly opened process, but not so much, just around 670MB).
In Google Chrome it looks the same. To reproduce it open browsers Task Manager (Shift+Esc) and in Business Process Copy, Paste, Move, Delete subprocesses several times. - **OK, can you check with details this point**

@romartin can you review it as well, now the PR is updated.

Thanks you guys.

tiagodolphine commented:
jenkins execute full downstream build

tiagodolphine commented:
Jenkins please retest this

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/command/CloneCanvasNodeCommand.java:
```diff
@@ -0,0 +1,115 @@
+            });
+
+            //process children children connectors and dock
+            childrenTraverseProcessor.consume(graph, getCandidate(), node -> {
```
why not consuming the graph nodes only once? I mean it seems it's exactly the same code just a few lines above

tiagodolphine replied:
@romartin the point here is, to add the cloned connectors we must have the cloned nodes already in the canvas, if we try to add clone connector and the source/target are not there than we got an exception. So to solve this I decided to first have all the cloned nodes and after that add the cloned connectors, what do you think? have any other idea?

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/canvas/command/CloneCanvasNodeCommand.java:
```diff
@@ -0,0 +1,115 @@
+    public CloneCanvasNodeCommand(Node parent, Node candidate, String shapeSetId) {
+        super(parent, candidate, shapeSetId);
+
+        this.childrenTraverseProcessor = new ChildrenTraverseProcessorImpl(new TreeWalkTraverseProcessorImpl());
```
probably better injecting the processor instance as a managed CDI bean rather than using 'new' (we can specialize it if necessary at some point). In fact the command instances are not managed beans, so you cannot directly inject here, but they're being created in the `CanvasCommandFactory`, which is a managed bean, this way you can add the processor as a constructor argument and pass from the factory bean a new processor instance.... see for example `DrawCanvasCommand`

tiagodolphine replied:
@romartin yeah this is better, in this case I need to inject the CanvasCommandFactory on the command, I will do that, just a little bit of code, but I think It will be better, too !

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-core-api/src/main/java/org/kie/workbench/common/stunner/core/graph/processing/traverse/content/ChildrenTraverseProcessor.java:
```diff
@@ -28,4 +31,6 @@
 
     ChildrenTraverseProcessor setRootUUID(final String rootUUID);
+
+    void consume(Graph graph, Node<?, ? extends Edge> parent, Consumer<Node<?, ? extends Edge>> nodeConsumer);
```
hehe yeah it sounds good for me this method, but in fact it does the same as if you use the callback for this processor, right? I mean both are doing the same... so IMO I prefer this new method, but sounds a bit confusing when injecting this type, which one should I use? By looking at the impl you realize that both are the same but the interface sounds a bit confusing for me... just as a comment, it's my opinion :)
This type is used in several places I think, so probably not a good idea trying to refactor it at this point and just keep one method, right? 
Well just a minor comment, you can add some docs too for making this more clear when looking at the interface too, and we can do some refactors later, that's fine for me too :)

tiagodolphine replied:
@romartin yeah, the idea here was to keep the code cleaner, using the callback I was repeating myself on a lot of places, this was the way I thought to keep the code cleaner but in fact we can make the same operation using both methods, this method I created is more a `helper` to have less code in some places, I can remove and use only the callback or move to other class, what do you think?

romartin commented:
Hey @tiagodolphine ,

Great job! Not much to say about the code, looks great in general, and tests too, just added a few review comments.

On the other hand found a bug and have some comments:

- Bug on Case Management - when pasting a task it ends up in a wrong layout (see screenshot). I think it's just probably missing some client side refresh call after the paste... as if you do some other action the task gets placed in the right location.

![screenshot from 2017-11-17 21-41-14](https://user-images.githubusercontent.com/4602417/32970870-f68a8356-cbea-11e7-8b8a-78a24c22fbbc.png)

- Missing to be able to paste same copied element several times - really used from the end user perspective. Probably just applies when copying, not cutting. I understand it's just probably not clearing the clipboard after the copy/paste.. well you have the detail, sounds easy to do?

- Finally I think we're missing some top level component yet. The `ClipboardControl` is being injected in the `Copy/Cut/PasteSessionCommand`'s, and actually these are used by the toolbar or the editor's buttons, so at this point you're injecting your beans and listening for the keybard events etc, but if someone just injects a new Stunner client session and the diagram presenter (which has no toolbar)? Welll just take it a look, there exist a type called `CanvasControl` which has different implementations, and the idea of these controls is that once you inject a `ClientSession` it binds the different controls to that canvas/handler, same way if the session is paused (changing to another editor) or destroyed it does some stuff too. This way the `ClipboardControl` becomes a `@Dependent` bean managed by this top level control class, which can inject at same time the managed instances for cut/copy/pasting logic. This way too, although not including the toolbar/editor buttons, the keyboard will still keep listening to `Ctrl+C/X/V`. So basically the idea is not binding these features to the buttons, is about binding to the stunner session level. If you look the `SessionCommand'`s you implemented are already bind to the current `ClientFullSession` instance, so you would be able to consume the new top level canvas control instance from the buttons and just call cut, copy, paste methods (notice even there is no need to inject the clipboard into the buttons this way..)

Well last point sounds big change but it is not hehe, just take it a look,  the other features (drag, resize, selection, etc) are built like this as well, the idea on the buttons is that they just use (canvas controls bind the session), if possible. WDYT?

Anyway really good job, happy with this, just let's try to do last cleans and that's it :)

Thanks!!

tiagodolphine commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/toolbar/command/PasteToolbarCommand.java:
```diff
@@ -0,0 +1,54 @@
+import org.kie.workbench.common.stunner.core.client.session.ClientFullSession;
+import org.kie.workbench.common.stunner.core.client.session.command.impl.PasteSelectionSessionCommand;
+
+// TODO: I18n.
```
@romartin ooh ok, this was a copy/paste from other command but it is better to handle the i18n for sure, I will do that.

tiagodolphine commented:
@romartin good points and really makes sense, I will work on all of them and let you know when the PR is updated. Thanks for your comments and review!

hasys commented:
Jenkins execute full downstream build

tiagodolphine commented:
Hey @romartin I updated the PR with the fix and improvements regarding your comments and review, the changes are on this commig https://github.com/kiegroup/kie-wb-common/pull/1286/commits/6915868fd5cabc8f0848ea05aa746a82eb60e65e

c/c @hasys

tiagodolphine commented:
Jenkins execute full downstream build

tiagodolphine commented:
Jenkins please retest this

tiagodolphine commented:
Jenkins please retest this

romartin commented:
Jenkins please retest this

tiagodolphine commented:
@romartin now when changing between diagrams, the clipboard control is cleared, I need to implement using Events, see https://github.com/kiegroup/kie-wb-common/pull/1286/commits/fe9baa8deb79cd203c37fe31a1089dd8dda5892e because the unbind was not being called when changing multiple times the diagrams.

tiagodolphine commented:
Jenkins please retest this

tiagodolphine commented:
Jenkins execute full downstream build

tiagodolphine commented:
Hi @hasys  can you review again to check if the point you raised were solved.
Thanks !

tiagodolphine commented:
Jenkins execute full downstream build

romartin commented:
@hasys same here, WAR files are generated even the build is failed, can you please test this and give us feedback asap? Once you're happy I'll merge.
CC @tiagodolphine 
Thanks!

tiagodolphine commented:
Jenkins execute full downstream build

romartin commented:
@hasys please remember to test this one, it's here for long time and I would like to merge ASAP... thanks!

hasys commented:
Hi @tiagodolphine,

looks good, thank you, but I still see some issues there:

1. Not possible to copy and paste Swimlane with Embedded Sub-Process inside (error dialog will be shown), process can be messed.
    * **Steps to reproduce:** 
    * Create Swimlane with non empty Embedded Sub-Process. 
    * Copy and paste swimlane. 
    * Close error dialog. 
    * Move swimlane or perform any other action. 
    * Save process and reopen it.
    * **Actual Result:**
    * All elements will be duplicated on the canvas
    * **Expect Result:**
    * Process will have the same amount of elements after reopen as before
    * Copy and paste of Swimlane with Sub-Process will be performed without any issues.
2. Some times sequence flows are under Sub-process after copy-paste
    * **Steps To Reproduce:**
    * Create Swimlane with non empty Sub-Process
    * Copy and paste Sub-Process
    * Move new Sub-Process out of Swimlane
    * **Actual Result:**
    * All sequence flows will be under Sub-Process and will be “invisible” for user
    * **Expected Result:**
    * Sequence flows always will be above Sub-Process
3. (Probably unrelated) You can make designer unusable
    * **Steps to Reproduce:**
    * Add Sub-Process to canvas
    * Add User task to Canvas
    * Add Boundary event to User task
    * Move User task with boundary task to Sub-Process
    * **Actual Result:**
    * Error message will be shown. When I closed it and played with Sub-Process little bit more Process started to show me error message on any mouse move.
    * **Expected result:**
    * Task will be added to Sub-Process without any problems.
4. Not possible to copy and paste between processes

tiagodolphine commented:
@hasys thanks for the review and the issues you've found. 
1) fixed
2) fixed
3) created a new JIRA https://issues.jboss.org/browse/JBPM-6682
4) for now we will not allow copy/paste between different diagrams, but we will handle that in the future,  was that what you mean? 

Thanks

c/c @romartin

tiagodolphine commented:
Jenkins please retest this

hasys commented:
Jenkins execute full downstream build

tiagodolphine commented:
Jenkins please retest this

romartin commented:
hey  @hasys , great job here as well :) 
It seems @tiagodolphine has fixed your comments and artifacts have been generated in the jenkins job, please can you give it a try and verify the fixes? I'll merge anyway in case the builds are failing due to the already known test cases... thanks!!

tiagodolphine commented:
Jenkins execute full downstream build

tiagodolphine commented:
Hi @hasys yeah as @romartin said the PR is updated with the fixes to the issues you found, the full downstream is not green but you can use the generated artifacts on your tests, since the errors were caused by tests not related to the PR.
Thanks !!

hasys commented:
Hi @tiagodolphine, thank you! Sorry, but I found something more:
1. First issue from last list fixed only partially:
   - **Steps to reproduce:**
   - Connect subprocess in lane from last steps to reproduce with start and end nodes outside of the Lane.
2. When you copy and paste resized Task/Subprocess/Lane after save and reopen of business process pasted element size will be reset to default one.
   - **Steps to reproduce:**
   - Add to the process a task, resize it
   - Copy and paste task, save and reopen business process
   - **Actual result:**
   - Pasted task size will be reset to default values.

tiagodolphine commented:
@hasys thank you for the new review and the issues you found, I already fixed them and push the new commit to the branch, so if you have a chance to check it again :)

Thanks in advance !
c/c @romartin

tiagodolphine commented:
Jenkins execute full downstream build

romartin commented:
Hey @hasys @tiagodolphine , I'll merge this anyway, it's taking too long, please can you open a ticket for the bug found? Thanks!

romartin commented:
hey @hasys @tiagodolphine I have just created this JIRA ticket in order to keep these issues found for lanes (in other PRs too) under control -> see [JBPM-6701](https://issues.jboss.org/browse/JBPM-6701)

tiagodolphine commented:
@romartin ok thanks to open the new JIRA ticket and for the merging. 
@hasys good catch, I didn't assume that use case on the copy/paste but we can work on that on the new ticket.

Thanks !
